src/vm/threads.cpp

```diff
@@ -6009,6 +6042,80 @@ void ThreadStore::TransferStartedThread(Thread *thread, BOOL bRequiresTSL)
+{
+    // Although all increments and decrements are usually done inside a lock, that is not sufficient to synchronize with a
+    // background GC thread resetting this value, hence the interlocked operation. Ignore overflow; overflow would likely never
+    // happen, the count is treated as unsigned, and nothing bad would happen if it were to overflow.
+    SIZE_T count = static_cast<SIZE_T>(FastInterlockIncrement(&m_DeadThreadCountForGCTrigger));
+
+    SIZE_T countThreshold = static_cast<SIZE_T>(s_DeadThreadCountThresholdForGCTrigger);
+    if (count < countThreshold || countThreshold == 0)
+    {
+        return;
+    }
+
+    IGCHeap *gcHeap = GCHeapUtilities::GetGCHeap();
+    SIZE_T millisecondsSinceLastFinalizingGC =
+        gcHeap->GetNow() - gcHeap->GetLastGCStartTime(gcHeap->GetMinFinalizingGeneration());
+    if (millisecondsSinceLastFinalizingGC < s_DeadThreadGCTriggerPeriodMilliseconds)
+    {
+        return;
+    }
+
+    if (!g_fEEStarted)
+    {
+        return;
+    }
+
+    // The dead thread count used for triggering a GC is tracked separately from the actual dead thread count so that a dead
+    // thread does not contribute to triggering a GC more than once. The count exceeds a certain threshold and a finalizing GC
+    // has not occurred in a certain duration, so reset the count and trigger a GC now to finalize unreachable dead threads.
+    // The GC is triggered on the finalizer thread since it's not safe to trigger it on DLL_THREAD_DETACH.
+    m_DeadThreadCountForGCTrigger = 0;
+    m_TriggerFinalizingGC = true;
+    FinalizerThread::EnableFinalization();
+}
+
+void ThreadStore::DecrementDeadThreadCountForGCTrigger()
+{
+    // Although all increments and decrements are usually done inside a lock, that is not sufficient to synchronize with a
+    // background GC thread resetting this value, hence the interlocked operation.
+    if (FastInterlockDecrement(&m_DeadThreadCountForGCTrigger) < 0)
+    {
+        FastInterlockExchange(&m_DeadThreadCountForGCTrigger, 0);
+    }
+}
+
+void ThreadStore::OnFinalizingGCStarted()
+{
+    // A dead thread may contribute to triggering at most once. After a finalizing GC occurs, if some dead thread objects are
+    // still reachable due to references to the thread objects, they will not contribute to triggering a GC again. Synchronize
+    // the store with increment/decrement operations occurring on different threads, and make the change visible to other
+    // threads in order to prevent unnecessary GC triggers.
+    FastInterlockExchange(&m_DeadThreadCountForGCTrigger, 0);
+}
+
+bool ThreadStore::ShouldTriggerFinalizingGC()
+{
+    return m_TriggerFinalizingGC;
+}
+
+void ThreadStore::TriggerFinalizingGCIfNecessary()
+{
+    if (!m_TriggerFinalizingGC)
+    {
+        return;
+    }
+    m_TriggerFinalizingGC = false;
+
+    IGCHeap *gcHeap = GCHeapUtilities::GetGCHeap();
+    gcHeap->GarbageCollect(gcHeap->GetMinFinalizingGeneration(), FALSE, collection_non_blocking);
```

Comment:
Ok, I think this is different from the scenario I understood when we talked. The scenario I understood was the threads that you created before were all dead (except the current one) - nothing holding onto them - if something is holding onto them it means they are useful in some way and if that's the case, some work is in progress and GCs will probably be triggered anyway. But it sounds like this is not the case because you worry about a little bit of work is going on that's enough to trigger gen1's. I am hesitant to have it trigger gen2 GCs because then we might be back to square one where it's triggering too many gen2 GCs so we have to be careful about triggering gen2's. You could call WhichGeneration and see if the percentage of gen2 objects is high enough, then trigger a gen2 but I would experiment with this carefully.